engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerService.java

```diff
@@ -0,0 +1,268 @@
+import com.odysseusinc.arachne.executionengine.execution.Overseer;
+import com.odysseusinc.arachne.executionengine.execution.sql.SQLService;
+import com.odysseusinc.arachne.executionengine.model.descriptor.DescriptorBundle;
+import com.odysseusinc.datasourcemanager.krblogin.KrbConfig;
+import java.time.Instant;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.Map;
+import java.util.Arrays;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.BiConsumer;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.scheduling.concurrent.ThreadPoolTaskExecutor;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+@Slf4j
+public class DockerService extends RService {
+    @Autowired
+    @Qualifier("analysisTaskExecutor")
+    private ThreadPoolTaskExecutor executor;
+    @Autowired
+    private DockerProperties dockerProperties;
+
+    private static final String DOCKER_ENV_APP_DEBUG = "RUNTIME_ENV_APP_DEBUG";
+    private static final String DOCKER_ENV_MEMORY_LIMIT_KEY = "RUNTIME_ENV_DOCKER_MEMORY_LIMIT";
+    private static final String DOCKER_ENV_CPU_LIMIT_KEY = "RUNTIME_ENV_DOCKER_CPU_LIMIT";
+    private static final String DOCKER_ENV_MEMORY_LIMIT_VALUE = "512m";
+    private static final String DOCKER_ENV_CPU_LIMIT_VALUE = "0.5";
+
+    @Override
+    public Overseer analyze(
+            AnalysisSyncRequestDTO analysis, File file, DescriptorBundle bundle,
+            KrbConfig krbConfig, BiConsumer<String, String> callback, Integer updateInterval) {
+        Instant started = Instant.now();
+        StringBuffer stdout = new StringBuffer();
+        BiConsumer<String, String> callbackWithStdOut = callback.andThen((stage, out) -> {
+            stdout.append(out);
+        });
+
+        CompletableFuture<ExecutionOutcome> future = CompletableFuture.supplyAsync(() -> {
+            try (DockerClient client = dockerClient()) {
+                String image = analysis.getDockerImage();
+                pullImageIfNotExists(image);
+                log.info("Creating container for image: {} and analysis Id: {}", image, analysis.getId());
+                CreateContainerResponse container = client.createContainerCmd(image)
+                        .withStdinOpen(true)
+                        .withTty(true)
+                        .exec();
+                log.info("Container created for image {} with container id {}", image, container.getId());
+                String containerId = container.getId();
+                runContainer(client, containerId, file, analysis, callbackWithStdOut);
+                remove(client, containerId);
```

Comment:
If exception occurs in runContainer, the container will not be removed. Is this intentional?

Reply:
Thanks. This is not intentional. I moved this code to the finally block inside `runContainer` to ensure resources are always cleaned up, even in case of exceptions but it could be useful for debugging purposes.